File: d02/src/step1.py
import re

def main():
    count = 0
    c = re.compile(r'^(\d+)-(\d+) (\w): (\w+)$')
    b = lambda a, b, c, d: (1 if d[int(a) - 1] == c else 0) + (1 if d[int(b) - 1] == c else 0)
    with open('./input.txt') as fd:
        for line in fd.readlines():
            m = c.match(line)
            if m:
                try:
                    if b(m[1], m[2], m[3], m[4]) == 1:
                        count += 1
                except:
                    pass
    print(count)

if __name__ == '__main__':
    main()

File: 2020/d02/src/step1.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let sl = |ip: &String| -> Option<bool> {
        let dash = ip.chars().position(|c| c == '-').unwrap();
        let space = ip.chars().position(|c| c == ' ').unwrap();
        let mini = ip.get(0..dash).unwrap().parse::<usize>().unwrap();
        let maxi = ip.get(dash + 1..space).unwrap().parse::<usize>().unwrap();
        let to_find = ip.chars().nth(space + 1).unwrap();
        let count = ip.chars().skip(space + 3).filter(|c| *c == to_find).count();
        let cond = count >= mini && count <= maxi;
        return Some(cond);
    };

    let mut sledge = 0;

    // File hosts must exist in current path before this produces output
    if let Ok(lines) = read_lines("./input.txt") {
        // Consumes the iterator, returns an (Optional) String
        for line in lines {
            if sl(&line.unwrap()) == Some(true) {
                sledge += 1;
            }
        }
        println!("Match {}", sledge);
    } else {
        println!("Sucked")
    }
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
